bundle/src/main/java/com/adobe/acs/commons/packaging/impl/AssetPackagerServletImpl.java

```diff
@@ -0,0 +1,281 @@
+                    }
+                }
+                paths.add(0, pageFilter);
+            }
+
+            doPackaging(request, response, preview, properties, paths);
+        } catch (RepositoryException ex) {
+            log.error("Repository error while creating Asset Package", ex);
+            response.getWriter().print(packageHelper.getErrorJSON(ex.getMessage()));
+        } catch (IOException ex) {
+            log.error("IO error while creating Asset Package", ex);
+            response.getWriter().print(packageHelper.getErrorJSON(ex.getMessage()));
+        }
+    }
+
+    /**
+     * Gets the properties saved to the Asset Packager Page's jcr:content node.
+     *
+     * @param request The request obj
+     * @return A ValueMap representing the properties
+     */
+    private ValueMap getProperties(final SlingHttpServletRequest request) {
+        if (request.getResource().getChild("configuration") == null) {
+            log.warn("Asset Packager Configuration node could not be found for: {}", request.getResource());
+            return new ValueMapDecorator(new HashMap<>());
+        } else {
+            return request.getResource().getChild("configuration").adaptTo(ValueMap.class);
+        }
+    }
+
+    /**
+     * Takes a list of Strings that should be valid patterns. If they are not then the exception is
+     * logged and the pattern is not added to the list of checked patterns.
+     *
+     * @param input The array of intended patterns
+     * @return The converted array of patterns
+     */
+    private List<Pattern> generatePatterns(String[] input) {
+        final List<Pattern> patterns = new ArrayList<>();
+        for (String item : input) {
+            try {
+                patterns.add(Pattern.compile(item));
+            } catch (PatternSyntaxException e) {
+                log.error("Pattern invalid, skipping. Pattern value: " + item);
+            }
+        }
+        return patterns;
+    }
+
+    /**
+     * Recursively iterate over the parent path specified in the configuration to aggregate all the
+     * String or String Array property values that are referencing a path in the DAM.
+     *
+     * @param resourceResolver Resolver from the current Request
+     * @param pagePath         The page path
+     * @return The full list of filter sets
+     */
+    private List<PathFilterSet> findAssetPaths(final ResourceResolver resourceResolver,
+                                               final String pagePath) {
+
+        final List<PathFilterSet> filters = new ArrayList<>();
+        final Resource parentResource = resourceResolver.resolve(pagePath);
+        if (isExcluded(this.pageExclusionPatterns, pagePath)) {
+            final PageManager pageManager = resourceResolver.adaptTo(PageManager.class);
+            Page page = pageManager.getContainingPage(parentResource);
+            if (!excludedPages.contains(page.getPath())) {
```

Comment:
Might simplify all of the code to use `Set` (even `LinkedHashSet` to keep order if you wish) in order to more easily avoid duplicates in all of the different collections we're playing with in this code.

Reply:
For example, if `excludedPages` is a Set, these 3 lines can be reduced to 1

Reply:
Don't feel too bad.  I've been saying the same thing for 20 years.  It's a journey.  😄